fix(material-experimental/mdc-typography): update mdc components typography to match spec

mmalerba opened the pull request:
This PR contains several change that bring the MDC-based components in line with the spec w.r.t typography.
1. It adds a new typography config generation function that creates a config based on values from MDC itself. This results in changes to some typography levels, most noticeable the `body-1` and `body-2` levels. Since the config is based on MDC's values it contains `rem` measurements rather than `px` measurements that Angular Material as traditionally used. In this PR the new config is used in the dev-app **only for the MDC-based components**. The MDC-based config is currently not compatible with the old components, but it will be made compatible in a followup PR.
2. Switches all the MDC components to read 2018 typography levels if they are available in the config. If they are not available, the 2014 levels are mapped to 2018 levels so the MDC components can consume them.
3. Corrects several places where the wrong typography level was being used (`body-1` looks very similar to `subtitle-1` which lead to them being used interchangeably).
4. Corrects some sizing styles for the MDC-based form-field so it looks right with the new typography levels. 

Demo: https://mmalerba-demo.web.app/mdc-autocomplete

CARETAKER NOTE: mmalerba will handle syncing

crisbeto commented on src/material-experimental/mdc-autocomplete/_autocomplete-theme.scss:
```diff
@@ -19,6 +20,12 @@
+      // MDC uses the `subtitle1` level for list items, but the spec shows `body1` as the correct
+      // level.
+      .mat-mdc-option {
+        @include mdc-typography(body1, $query: $mat-typography-styles-query);
```
These seem to be the same between autocomplete and select. Maybe we should do it once inside the option theme instead?

mmalerba replied:
I didn't see an existing mixin for option typography

crisbeto replied:
There's `mat-mdc-option-typography`. Also now that I'm thinking about it, we likely need the same typography change for `mat-optgroup` since it uses the same size as the option.

mmalerba replied:
Oh hmm, don't know how I missed that, I'll move it

crisbeto replied:
The groups are set up to look like options, except that they aren't clickable, so the same styles can be applied.

mmalerba replied:
I think they're currently set to be subtitle1 which makes sense, after the typography levels are updated it'll wind up looking nearly identical to the options (except for the letter spacing) https://material.io/design/typography/the-type-system.html#type-scale

crisbeto replied:
Funnily enough, the link you just sent has an example of an option group which is somewhat different than what we have. Anyway, my comment was mostly to raise awareness that we might want to change the option groups as well, I don't have an opinion on what they should be changed to.
![The_type_system_-_Material_Design_-_Google_Chrome_2021-01-26_23-33-08](https://user-images.githubusercontent.com/4450522/105914669-eb74e200-602e-11eb-9107-5b18fbabba89.png)

mmalerba replied:
I created a PR based on this one that uses a config created from MDC's typography settings: https://github.com/angular/components/pull/21709 That should give an idea of how things should eventually end up looking. Though looking at it, the form-field appears to need a little more work

mmalerba replied:
It looks like they're using the `overline` level for the option group, we may want to do that too but I'll leave it for another PR.

mmalerba commented:
I discussed with Jeremy and we decided its best to roll the function to create a typography config from MDC's values into this PR. I've updated it and added a demo of the latest code in the PR description.

jelbourn commented:
Does the mdc-paginator look weird?
![image](https://user-images.githubusercontent.com/838736/106198787-5591a200-6169-11eb-8df8-c48a3a4803a3.png)

mmalerba commented:
Yeah it does look a little weird. I think the paginator is just applying the dense form-field styles (which does not change the font size). I can try adding some overrides though

mmalerba commented:
added an override for the paginator and updated the demo

jelbourn commented on src/material-experimental/mdc-helpers/_mdc-helpers.scss:
```diff
@@ -51,9 +52,45 @@ $mat-typography-level-mappings: (
   )
 );
 
+// Mappings between Angular Material and MDC's typography levels.
```
Could you expand this comment to be explicit about the direction of the mapping? I.e.
```scss
// Maps ___ to ____. 
```
Might also be good to touch here on how this is intended to be used

jelbourn commented on src/material-experimental/mdc-typography/_all-typography.scss:
```diff
@@ -1,5 +1,52 @@
+@import '@material/typography/variables.import';
 @import '../mdc-theming/all-theme';
 
+@function mat-mdc-typography-config(
```
Add a SassDoc for this function?

(I know we haven't done this much in the past, but I'd like to start)

mmalerba commented on src/material-experimental/mdc-core/option/_option-theme.scss:
```diff
@@ -43,6 +44,14 @@
+    // MDC uses the `subtitle1` level for list items, but the spec shows `body1` as the correct
+    // level. Class is repeated for increased specificity.
+    .mat-mdc-option {
+      @include mdc-typography(body1, $query: $mat-typography-styles-query);
```
We have an internal one filed against our gmat components - the difference is almost imperceptible with the external spec, just the letter spacing is different. I haven't opened an bug against MDC yet.

mmalerba commented on src/material-experimental/mdc-form-field/_form-field-subscript.scss:
```diff
@@ -8,16 +8,34 @@
+    position: absolute;
+    top: 0;
+    left: 0;
+    right: 0;
```
Yes, the idea is to behave like the non-MDC form-field where the user can add multiple error/hint or long ones that wrap, but they'll have to reserve space for it themselves. We only reserve space in the layout for one line worth of message.

jelbourn replied:
We should maybe talk about this- I think we should change our behavior over to what mdc does, but I know that would be pretty breaking, so we should think about doing a follow-up that lets users switch.

mmalerba commented on src/material-experimental/mdc-form-field/_mdc-text-field-structure-overrides.scss:
```diff
@@ -9,15 +9,18 @@
   }
 
+  // In order to ensure proper alignment of the floating label, we reset its line-height.
+  // The line-height is not important as the element is absolutely positioned and only has one line
```
not really, it *does* affect the size of the element - its just that we absolutely position the element, so the size doesn't affect the rest of the layout, and its not important as far as the line spacing since labels are only one line anyways

jelbourn reviewed (approved):
LGTM
